### DISCUSSION TITLE
Is there a way to query only one element among descendants?

### DISCUSSION THREAD

**Discussion by mtskf** [renggli/dart-xml#115] [Q&A] [answered]
I'm looking for a way to query a first-matching element among descendants like .querySelector() in html.
`.findAllElements('...').first` queries all the matching elements and costly, and I need just only one.

I'd appreciate if you could advise! Thanks 😊

**Comment by renggli** [comment, marked as answer]
`findAllElements` returns a lazy `Iterable` (see [the documentation](https://github.com/renggli/dart-xml/blob/11f919e17dc3099ae267b5efb1daa8a6d7ce3952/lib/src/xml/navigation/find.dart#L41), there is no `List` with all elements created). The iterable doesn't [do anything](https://github.com/renggli/dart-xml/blob/11f919e17dc3099ae267b5efb1daa8a6d7ce3952/lib/src/xml/navigation/descendants.dart#L18) until you actually iterate over the elements.

When you call `first` on the iterator, it starts to traverse the tree until it finds the first matching element, aborts the traversal, and returns that element. Check out the Dart tutorial on [Iterable collections](https://dart.dev/codelabs/iterables) for other operations.

**Comment by mtskf** [comment]
@renggli 
Thank you! 👍
